src/core/server/plugins/plugins_service.ts

```diff
@@ -19,17 +19,20 @@
 
 import { Observable } from 'rxjs';
 import { filter, first, mergeMap, tap, toArray } from 'rxjs/operators';
-import { CoreService } from '../../types';
+import { CoreService, PluginName } from '../../types';
 import { CoreContext } from '../core_context';
 import { ElasticsearchServiceStart } from '../elasticsearch';
 import { Logger } from '../logging';
 import { discover, PluginDiscoveryError, PluginDiscoveryErrorType } from './discovery';
-import { Plugin, PluginName } from './plugin';
+import { DiscoveredPlugin, Plugin } from './plugin';
 import { PluginsConfig } from './plugins_config';
 import { PluginsSystem } from './plugins_system';
 
 /** @internal */
-export type PluginsServiceStart = Map<PluginName, unknown>;
+export interface PluginsServiceStart {
+  pluginStarts: Map<PluginName, unknown>;
+  discoveredPlugins: Map<PluginName, DiscoveredPlugin>;
```

Comment:
I'm not a big fan of the name `pluginStarts`, but I'm especially not a fan of the notion of plugin "discovery" leaking through in the name of this.

How about simply `contracts` and applying the pattern for all lifecycle functions even if `start` is the only current one that deals with discovery as well.

Is `discoveredPlugins` used for anything other than passing to the client? If not, how about restricting its contents to only plugins with UI and calling this `uiPlugins`? And carefully select only the values from the manifest that we explicitly need to access from the UI so we never accidentally expose things we don't need.

I think I understand the motivation for keeping this generic, but in the end the notion of `ui` specific plugins _is_ a plugin manifest concern, so I think it makes sense for the plugin service to determine what is and is not relevant here.

So the interface would become:

```ts
{
  contracts: Map<PluginName, unknown>;
  uiPlugins: Map<PluginName, DiscoveredPlugin>;
}
```

Reply:
> How about simply `contracts` and applying the pattern for all lifecycle functions even if start is the only current one that deals with discovery as well.

With the standardization around call all of these interfaces `MyPluginStart` instead of `MyPluginStartContract` I thought using the word "contract" might not be as clear. Personally, I prefer interface names that are more conventional nouns, so `contracts` makes sense to me.

> Is discoveredPlugins used for anything other than passing to the client? If not, how about restricting its contents to only plugins with UI and calling this uiPlugins? And carefully select only the values from the manifest that we explicitly need to access from the UI so we never accidentally expose things we don't need.

This was the dilemma I was having up above: whether or not to make this UI specific. If we're fine with that, then yes I agree on limiting the scope of what information is leaked through this interface. Currently, I only plan to use this data in the optimizer and for injected metadata.